src/backend/distributed/master/master_delete_protocol.c

```diff
@@ -381,47 +391,70 @@ DropShards(Oid relationId, char *schemaName, char *relationName,
 		CoordinatedTransactionUse2PC();
 	}
 
-	ShardInterval *shardInterval = NULL;
-	foreach_ptr(shardInterval, deletableShardIntervalList)
+	List *dropTaskList = DropTaskList(relationId, schemaName, relationName,
+									  deletableShardIntervalList);
+
```

Comment:
The flow of `DropShards ` is different than all other places that we do local execution. We do local execution per placement here -- even not per task. Can we use `ExtractLocalAndRemoteTasks` trick and execute all local drops within the same list? Similarly, with the remote tasks, go over them in a separate loop.

I can see that we do `DeleteShardPlacementRow` / `DeleteShardRow`, and you probably preferred not to change the structure. However, everything is happening in a transaction, so any failure would rollback whole  transaction. So should be OK to move a separate function/loop at the end of all drop tasks done

Reply:
I was also considering such a change, now trying it and will update this comment

Reply:
I wonder what we will gain by separating the loop into pieces, only the code in `DropShards` will grow ?

Reply:
Actually I'm more inclined to change the execution logic in DropShards to the ones in adaptive executor (calling `ExecuteUtilityTaskListWithoutResults`) after this task, in a seperate pr

Reply:
> I wonder what we will gain by separating the loop into pieces, only the code in DropShards will grow ?

It'll grow, and probably become slightly slower. But, it'd be (a) easier to follow (b) resemble to the existing codes that do the same thing

Reply:
> It'll grow, and probably become slightly slower. But, it'd be (a) easier to follow (b) resemble to the existing codes that do the same thing

Looking into this further, I agree with @onurctirtir, it'd make things more complicated if we try to split the loops. So, I'm fine with keeping it